### ISSUE TITLE
[Bug]: Pickled figure artists can no longer be picked

### ISSUE THREAD

**Issue by moux233** [matplotlib/matplotlib#21802]
### Bug summary

I usually pickle my figures to disk, and they routinely include some callbacks.
When I load my figures from disk, callbacks need to be re-activated using fig.canvas.mpl_connect.
Some callbacks turn back on, such as "key_press_event" and "button_press_event".
However, I can not get "pick_event" to work.
The code below is based on matplotlib's legend picking example.
Note the assertion showing that the pickled legend line has a correct picker property.

### Code for reproduction

```python
import io
import pickle

import numpy as np
import matplotlib.pyplot as plt


# %% creating original figure

t = np.linspace(0, 1)
y1 = 2 * np.sin(2 * np.pi * t)
y2 = 4 * np.sin(2 * np.pi * 2 * t)

fig, ax = plt.subplots()
ax.set_title('Click on legend line to toggle line on/off')
ax.plot(t, y1, lw=2, label='1 Hz')
ax.plot(t, y2, lw=2, label='2 Hz')
leg = ax.legend(fancybox=True, shadow=True)

lines = ax.lines
lined = {}  # Will map legend lines to original lines.
for legline, origline in zip(leg.get_lines(), lines):
    legline.set_picker(5)  # Enable picking on the legend line.
    lined[legline] = origline


def on_pick(event):
    print("on pick")
    # On the pick event, find the original line corresponding to the legend
    # proxy line, and toggle its visibility.
    legline = event.artist
    origline = lined[legline]
    visible = not origline.get_visible()
    origline.set_visible(visible)
    # Change the alpha on the line in the legend so we can see what lines
    # have been toggled.
    legline.set_alpha(1.0 if visible else 0.2)
    fig.canvas.draw()


def on_click(event):
    print("on click")


fig.canvas.mpl_connect('pick_event', on_pick)
fig.canvas.mpl_connect("button_press_event", on_click)


# %% pickling

buf = io.BytesIO()
pickle.dump(fig, buf)
buf.seek(0)
fig2 = pickle.load(buf)
assert fig2.get_axes()[0].get_legend().get_lines()[0].get_picker() == 5
fig2.canvas.mpl_connect('pick_event', on_pick)  # does not work :(
fig2.canvas.mpl_connect("button_press_event", on_click)  # works all right

plt.show()
```

### Actual outcome

2 identical figures.
The original figure behaves toggles line artists on and off when the legend is picked. It also prints "on click" and "on pick" messages in the console.
The second figure gets the "on click" event, but not the "on pick".

### Expected outcome

Same behaviour for both figures.

### Additional information

Same bug within Spyder or at the console.
Thanks for your help.

### Operating system

Windows 10

### Matplotlib Version

3.5.0

### Matplotlib Backend

Qt5Agg

### Python version

3.9.7

### Installation

pip

**Comment by QuLogic** [comment]
Did this use to work in a previous version?

**Comment by moux233** [comment]
I just checked that I have the same bug with matplotlib 3.4.0.
I don't think it ever worked, I simply did not notice it before.

**Comment by QuLogic** [comment]
On current `v3.7.x`, the `on_pick` callback is triggered, but crashes, since the `lined` dictionary only contains lines from the pre-pickle figure. Also, `on_pick` uses the original figure to refresh on change, so nothing appears to happen even when it's working. If I change the code slightly to the following, then all works:
```python
import io
import pickle

import numpy as np
import matplotlib.pyplot as plt


# %% creating original figure

t = np.linspace(0, 1)
y1 = 2 * np.sin(2 * np.pi * t)
y2 = 4 * np.sin(2 * np.pi * 2 * t)

fig, ax = plt.subplots()
ax.set_title('Click on legend line to toggle line on/off')
ax.plot(t, y1, lw=2, label='1 Hz')
ax.plot(t, y2, lw=2, label='2 Hz')
leg = ax.legend(fancybox=True, shadow=True)

lines = ax.lines
lined = {}  # Will map legend lines to original lines.
for legline, origline in zip(leg.get_lines(), lines):
    legline.set_picker(5)  # Enable picking on the legend line.
    lined[legline] = origline


def on_pick(event):
    print("on pick")
    # On the pick event, find the original line corresponding to the legend
    # proxy line, and toggle its visibility.
    legline = event.artist
    origline = lined[legline]
    visible = not origline.get_visible()
    origline.set_visible(visible)
    # Change the alpha on the line in the legend so we can see what lines
    # have been toggled.
    legline.set_alpha(1.0 if visible else 0.2)
    event.canvas.draw()


def on_click(event):
    print("on click")


fig.canvas.mpl_connect('pick_event', on_pick)
fig.canvas.mpl_connect("button_press_event", on_click)


# %% pickling

buf = io.BytesIO()
pickle.dump(fig, buf)
buf.seek(0)
fig2 = pickle.load(buf)
ax2 = fig2.get_axes()[0]
leg2 = ax2.get_legend()
fig2.canvas.mpl_connect('pick_event', on_pick)
fig2.canvas.mpl_connect('button_press_event', on_click)
for legline, origline in zip(leg2.get_lines(), ax2.lines):
    assert legline.get_picker() == 5
    lined[legline] = origline

plt.show()
```

**Comment by QuLogic** [comment]
Actually, with those changes, it works with 3.6.2 as well. Probably this was fixed by #23448.
